sktime/classification/deep_learning/base.py

```diff
@@ -0,0 +1,119 @@
+because we can generalise tags, _predict and _predict_proba
+"""
+__author__ = ["James-Large", "ABostrom", "TonyBagnall"]
+__all__ = ["BaseDeepClassifier"]
+
+from abc import ABC, abstractmethod
+
+import numpy as np
+from sklearn.preprocessing import LabelEncoder, OneHotEncoder
+from sklearn.utils import check_random_state
+
+from sktime.classification.base import BaseClassifier
+
+
+class BaseDeepClassifier(BaseClassifier, ABC):
+    """Abstract base class for deep learning time series classifiers.
+
+    The base classifier provides a deep learning default method for
+    _predict and _predict_proba, and provides a new abstract method for building a
+    model.
+
+    Parameters
+    ----------
+    batch_size : int, default = 40
+        training batch size for the model
+
+    Arguments
+    ---------
+    self.model = None
+
+    """
+
+    _tags = {"X_inner_mtype": "numpy3D"}
+    # can probably handle multivariate??
+
+    def __init__(self, batch_size=40, random_state=None):
+        super(BaseDeepClassifier, self).__init__()
+
+        self.batch_size = batch_size
+        self.random_state = random_state
+        self.model_ = None
+
+    @abstractmethod
+    def build_model(self, input_shape, n_classes, **kwargs):
+        """Construct a compiled, un-trained, keras model that is ready for training.
+
+        Parameters
+        ----------
+        input_shape : tuple
+            The shape of the data fed into the input layer
+        n_classes: int
+            The number of classes, which shall become the size of the output
+            layer
+
+        Returns
+        -------
+        output : a compiled Keras Model
+        """
+        ...
+
+    def _predict(self, X, **kwargs):
+        probs = self._predict_proba(X, **kwargs)
+        rng = check_random_state(self.random_state)
+        return np.array(
+            [
+                self.classes_[int(rng.choice(np.flatnonzero(prob == prob.max())))]
+                for prob in probs
+            ]
+        )
+
+    def _predict_proba(self, X, **kwargs):
+        """Find probability estimates for each class for all cases in X.
+
+        Parameters
+        ----------
+        X : an np.ndarray of shape = (n_instances, n_dimensions, series_length)
+            The training input samples.         input_checks: boolean
+            whether to check the X parameter
+
+        Returns
+        -------
+        output : array of shape = [n_instances, n_classes] of probabilities
+        """
+        # Transpose to work correctly with keras
+        X = X.transpose((0, 2, 1))
+        probs = self.model_.predict(X, self.batch_size, **kwargs)
+
+        # check if binary classification
+        if probs.shape[1] == 1:
+            # first column is probability of class 0 and second is of class 1
+            probs = np.hstack([1 - probs, probs])
+
+        return probs
+
+    def convert_y(self, y, label_encoder=None, onehot_encoder=None):
```

Comment:
maybe call this `convert_y_to_keras` or similar?
to avoid confusion and accidental overrides